projects/train/train/data/datasets/base.py

```diff
@@ -0,0 +1,455 @@
+                if k not in ["dec", "psi", "phi"]
+            ]
+            parameters = {k: parameters[:, i] for i, k in enumerate(keys)}
+            strain, parameters = self.inject(
+                background, cross, plus, parameters
+            )
+
+        return strain, parameters
+
+    # ================================================ #
+    # Dataloaders used by lightning
+    # ================================================ #
+
+    def train_dataloader(self):
+        # if we only have one training file
+        # load it into memory and use InMemoryDataset
+        if len(self.train_fnames) == 1:
+            train_background = self.load_background(self.train_fnames)
+            dataset = InMemoryDataset(
+                train_background,
+                kernel_size=int(self.hparams.sample_rate * self.sample_length),
+                batch_size=self.hparams.batch_size,
+                coincident=False,
+                batches_per_epoch=self.hparams.batches_per_epoch,
+                shuffle=True,
+            )
+        else:
+            dataset = Hdf5TimeSeriesDataset(
+                self.train_fnames,
+                channels=self.hparams.ifos,
+                kernel_size=int(self.hparams.sample_rate * self.sample_length),
+                batch_size=self.hparams.batch_size,
+                batches_per_epoch=self.hparams.batches_per_epoch,
+                coincident=False,
+            )
+
+        pin_memory = isinstance(
+            self.trainer.accelerator, pl.accelerators.CUDAAccelerator
+        )
+        dataloader = torch.utils.data.DataLoader(
+            dataset, num_workers=self.num_workers, pin_memory=pin_memory
+        )
+        return dataloader
+
+    def val_dataloader(self):
+        # TODO: allow for multiple validation segment files
+
+        # offset the start of the validation background data
+        # by the device id to add more diversity in the validation set
+        _, rank = self.get_world_size_and_rank()
+
+        # build waveform dataloader
+        cross, plus = self.val_waveforms
+        waveform_dataset = torch.utils.data.TensorDataset(
+            cross, plus, self.val_parameters
+        )
+        waveform_dataloader = torch.utils.data.DataLoader(
+            waveform_dataset,
+            batch_size=self.val_batch_size,
+            shuffle=False,
+            pin_memory=False,
+        )
+
+        # build background dataloader
+        val_background = self.val_background[0][:, rank:]
+
+        background_dataset = InMemoryDataset(
+            val_background,
+            kernel_size=int(self.hparams.sample_rate * self.sample_length),
+            batch_size=self.val_batch_size,
+            batches_per_epoch=len(waveform_dataloader),
+            coincident=False,
+            shuffle=False,
+        )
+
+        background_dataloader = torch.utils.data.DataLoader(
+            background_dataset, pin_memory=False
+        )
+        return ZippedDataset(
```

Comment:
What is the reason behind returning both dataloaders for validation and testing?

Reply:
So for validation and testing I wanted to make sure that things are deterministic between each epoch so that our validation score is directly comparable. So, I generate the validation and testing waveforms once up front before training begins.

Then, the waveforms and background are loaded together and injected in the `on_after_batch_transfer` function